service/matching/taskListManager.go

```diff
@@ -587,12 +591,30 @@ func (c *taskListManagerImpl) getIsolationGroupForTask(ctx context.Context, task
 		}
 		group, err := c.partitioner.GetIsolationGroupByDomainID(ctx, taskInfo.DomainID, partitionConfig, pollerIsolationGroups)
 		if err != nil {
+			// For a sticky tasklist, return StickyUnavailableError to let it be added to the non-sticky tasklist.
+			if err == partition.ErrNoIsolationGroupsAvailable && c.taskListKind == types.TaskListKindSticky {
+				return "", _stickyPollerUnavailableError
+			}
+			// if we're unable to get the isolation group, log the error and fallback to no isolation
 			c.logger.Error("Failed to get isolation group from partition library", tag.WorkflowID(taskInfo.WorkflowID), tag.WorkflowRunID(taskInfo.RunID), tag.TaskID(taskInfo.TaskID), tag.Error(err))
-		} else {
-			return group
+			return "", nil
+		}
+		// For a sticky tasklist, it is possible that when an isolation group is undrained, the tasks from one workflow is reassigned
+		// to the isolation group undrained. If there is no poller from the isolation group, we should return StickyUnavailableError
+		// to let the task to be re-enqueued to the non-sticky tasklist. If there is poller, just return an empty isolation group, because
+		// there is at most one isolation group for sticky tasklist and we could just use empty isolation group for matching.
+		if c.taskListKind == types.TaskListKindSticky {
+			pollerIsolationGroups = c.pollerHistory.getPollerIsolationGroups(time.Now().Add(-time.Minute))
+			for _, pollerGroup := range pollerIsolationGroups {
+				if group == pollerGroup {
```

Comment:
How do we know this sticky poller is in an ok zone, it's not drained or something?

Reply:
I missed the earlier point about this already being filtered down to the correct isolation-group on line 592